jena-core/src/main/java/org/apache/jena/mem/GraphMemUsingHashMap.java

```diff
@@ -0,0 +1,774 @@
+ * Main differences between GraphMemUsingHashMap and GraphMem:
+ * - GraphMem uses its own hash map and triple bag implementations while GraphMemUsingHashMap uses the standard
+ *   HashMap<K,V> and ArrayList<T>.
+ * - GraphMemUsingHashMap optimizes find operations by
+ *   - implementing every possible permutation to avoid unnecessary repeated condition checks (Node.isConcrete)
+ *   - careful order of conditions to fail as fast as possible
+ * - GraphMemUsingHashMap has the Graph#stream operations implemented as real java streams considering the same
+ *   optimizations as the find operations and not wrapping iterators to streams.
+ * - GraphMemUsingHashMap optimizes memory usage by using Node.getIndexingValue().hashCode() as hash keys instead
+ *   of the Node.getIndexingValue() object itself. This is totally fine, because values are lists.
+ *
+ * Benchmarks show that:
+ * - adding triples is much faster than on GraphMem
+ * - for large graphs this implementation need less memory than GraphMem
+ * - find and contains operations are a bit faster than GraphMem
+ * - stream operations are faster than GraphMem and can be accelerated even more by appending .parallel()
+ *
+ * The ExtendedIterator<> returned by Graph#find calls supports .remove and .removeNext to make it fully compatible with the
+ * usages of GraphMem in the whole jena repository.
+ *
+ * Adding triples while iterating on a result is not supported, but it was probably not intentional that GraphMem
+ * supported this in some cases. The implementation of ModelExpansion.addDomainTypes relayed on this behaviour, but it
+ * has been fixed.
+ */
+public class GraphMemUsingHashMap extends GraphMemBase implements GraphWithPerform {
+
+    private static final int INITIAL_SIZE_FOR_ARRAY_LISTS = 2;
+
+    private final Map<Integer, ArrayList<Triple>> triplesBySubject = new HashMap<>();
+    private final Map<Integer, ArrayList<Triple>> triplesByPredicate = new HashMap<>();
+    private final Map<Integer, ArrayList<Triple>> triplesByObject = new HashMap<>();
+
+    public GraphMemUsingHashMap() {
+        super();
+    }
+
+    /**
+     * Subclasses over-ride this method to release any resources they no
+     * longer need once fully closed.
+     */
+    @Override
+    protected void destroy() {
+        this.triplesBySubject.clear();
+        this.triplesByPredicate.clear();
+        this.triplesByObject.clear();
+    }
+
+    /**
+     * Add a triple to the triple store. The default implementation throws an
+     * AddDeniedException; subclasses must override if they want to be able to
+     * add triples.
+     *
+     * @param t triple to add
+     */
+    @SuppressWarnings("java:S1199")
+    @Override
+    public void performAdd(Triple t) {
+        var sIdx = t.getSubject().getIndexingValue().hashCode();
+        var sList = this.triplesBySubject.get(sIdx);
+        if(sList != null) {
+            if(sList.contains(t)) {
```

Comment:
In the pathological case where every triple has the same subject this operation is going to get progressively less performant over time especially with large graphs

Yes this is a pathological case but these things crop up with depressing frequently in the real world.

I'd be interested in seeing some benchmarking that considers that situation, or variations thereof, e.g. small pool of subject nodes with lots of unique triples about them.

One possible optimisation here would be to get all three lists (subject, predicate and object) and do the `contains()` scan on the smallest list (or if any list is empty bypassing it entirely)

Reply:
Thank you very much for that hint!
While describing the benchmarks and results today, I discovered one test, where .contains() was slower than GraphMem. 
So I changed my test setup to make the problem more visible.
Your guess is right: It is because there were many triples sharing the same subject. 

I found a solution matching my samle graphs by using a sorted ArrayList when it has more than x (currently 50) entries.
They are sorted by an artificial key, combined from the Nodes .getIndexingValue().hashCode(). It seems to have almost no negative side effects so far. Though, that solution would not enhance .find() performance. I will see what I can do with that.
Unfortunately, I will need a few days to make the implementation pretty and write tests.

Reply:
How large should I think large graphs? 
Beginnig at about 17.000.000 triples (in the graph) the performance of binary search on array lists seam to loose against the GraphMem (only with positive contains() - negative contains() and other results are still fine)
-> I could try to find solutions also beating graphs largert than that but mabe not in all disciplines. Where should I stop?

For this setup:
- single "large" graph
- 2 iterations warmup for each implementation -> results are from 3rd iteration
- used http://wbsg.informatik.uni-mannheim.de/bizer/berlinsparqlbenchmark/spec/BenchmarkRules/index.html#datagenerator
 - java -cp lib/* benchmark.generator.Generator -pc 50000 -s ttl -ud

Results
bsbmtools-0.2\dataset.ttl - triples: 17424059
Reading graph to ArrayList: 00:01:33.741 using 3076 MB
GraphMem: 
- Graph.add() for all triples: 00:00:38.332
- additional memory used: 1830 MB
- for all triples: Graph.contains(): 00:00:02.389
- for same number of triples with generated random triples: Graph.contains():  00:00:48.626

GraphMemUsingHashMap:
- Graph.add() for all triples: 00:00:08.805
- additional memory used: 723 MB
- for all triples: Graph.contains(): 00:00:03.559
- for same number of triples with generated random triples: Graph.contains():  00:00:23.167

Any comments of my methology for this benchmark are welcome.
Details on other benchmarks will follow soon. The other graphs I used smaller and CGMES (Common Grid Model Exchange Standard) RDF graphs.

Reply:
> Where should I stop?

The test for merging this PR and replacing GraphMem isn't "must be better in every way on first release". We, the developer community, have to make judgement call as to whether, overall, it is better for users.

Reply:
I added a section "pathological case where every triple has the same subject" in the PR description. 
Including benchmarks and an alternative approach. (in another branch not yet equipped with tests for the additional classes).